### PR TITLE
Pokemon's info page

### PR DESCRIPTION
**Pull request by fabe85**
@samitsv 
@vivek-sethia 
@swathi-ssunder

### CODE REVIEW

File: controller/loadBasicPokemonDetails.js [hunk 1]
```diff
@@ -1,26 +1,35 @@
 'use strict';
 
 let fs = require('fs'),
-    basicPokemonDetail = require('../models/BasicPokemonDetail.js'),
+    basicPokemonDetail = require('../models/basicPokemonDetail.js'),
     pokemonListPath = require('../resources/json/pokemonlist.json');
```
Comment (swathi-ssunder) [inline controller/loadBasicPokemonDetails.js]: The file is currently called BasicPokemonDetail.js(with an uppercase B). So please revert this change or rename the file too if you wish.

Comment (fabe85) [inline controller/loadBasicPokemonDetails.js]: Sorry, now it should be right

File: controller/loadBasicPokemonDetails.js [hunk 2]
```diff
@@ -7,20 +7,29 @@ let fs = require('fs'),
...
+            /*if (female) {
+                gender = female;
+            } else {
+                gender = male;
+            }*/
+            //gender = value.gender.male;
             //add pokemonID and pokemonName to collection
-            basicPokemonDetail.create({pokemonID: pokemonID, name: pokemonName}, function (err, post) {
-                if (err) {
+            basicPokemonDetail.create({pokemonID: pokemonID, name: pokemonName, gender: {male:male,female:female}}, function(err, post){
```
Comment (sacdallago) [inline controller/loadBasicPokemonDetails.js]: Why would you store two keys for a binary attribuite? If someone is female she is not male. This is against NF (especially in NRDBS) and redundant :)
Also: what are you exactly storing? A string? You only need a boolean (which takes up much less space in memory)

Comment (sacdallago) [inline controller/loadBasicPokemonDetails.js]: Ok, ignore the second part

File: app/models/BasicPokemonDetail.js [hunk 3]
```diff
@@ -9,6 +9,182 @@ let BasicPokemonDetailSchema = new mongoose.Schema({
     },
     name: {
         type: String
+    },
+    gender: {
+        /*male: {
```
Comment (samitsv) [inline app/models/BasicPokemonDetail.js]: i would recommend make gender as a string, so that it will be like gender: "male" in the json, for category and move maybe array of strings rather than having all of the moves as a field with boolean in them

Comment (fabe85) [inline app/models/BasicPokemonDetail.js]: I also thought about that but boolean data are faster and easier to read. I mean the data are only relevant for the machine learning algorithms and instead of searching for strings which have to be brought in the correct form and where errors are more likely to appear when something is not correctly written the guys just have to execute some if statements. So I would make booleans everywhere it is possible.

**Comment by jonas-he** [comment]
I think the attributes here are more related to the info page instead of the machine learning, just sayin. And for the generation 1 pokemon (which are currently all that are in Pokemon Go) there is no gender except for the Nidoran family.

File: app/models/BasicPokemonDetail.js (hunk 3 above)
Comment (samitsv) [inline app/models/BasicPokemonDetail.js]: @swathi-ssunder, @vivek-sethia , @jonas-he what do you guys think?

Comment (vivek-sethia) [inline app/models/BasicPokemonDetail.js]: I think you can store gender as a boolean value, for e.g. 1 for male and 2 for female.

**Comment by goldbergtatyana** [comment]
You never know @jonas-he what kind of features will be important for machine learning (ML) later. Gender may not be the defining feature, but it might contribute a little to the prediction. The strategy for now is to get as much data as possible for ML groups, who then will do feature selection (determine the contribution of each feature you found for them). 

As for the gender we can show both venus mirror and mars shield & spear, as http://www.pokemon.com/us/pokedex/onix also do.

File: app/controllers/loadBasicPokemonDetails.js [hunk 4]
```diff
@@ -8,20 +8,20 @@ function loadBasicPokemonDetails() {
...
 
     let data = JSON.stringify(pokemonListPath, function(key,value) {
-        var pokemonID, pokemonName, male, female;
+        var pokemonID, pokemonName, male;
         if(key != 'name' && typeof value.name !== "undefined") {
             pokemonID = key;
             pokemonName = value.name;
             male = value.gender.male;
-            female = value.gender.female;
+            //female = value.gender.female;
```
Comment (sacdallago) [inline app/controllers/loadBasicPokemonDetails.js]: Another suggestion: git keeps the history of files. Commenting out old code is not a good practice, just delete it, you have revisions! :)

File: app/models/BasicPokemonDetail.js (hunk 3 above)
Comment (fabe85) [inline app/models/BasicPokemonDetail.js]: Well, if I say gender = 1,2 or, more concretely, gender: male = true, false, is a kind of gustus.

File: app/controllers/loadBasicPokemonDetails.js (hunk 4 above)
Comment (fabe85) [inline app/controllers/loadBasicPokemonDetails.js]: Ok thank you for the hint

**Comment by sacdallago** [comment]
Oh man. Sorry guys about making you change the sex thing, I didn't know there were hermaphrodite Pokémons.

I think the best idea is then to have a controlled dictionary (aka `String`, but limit it to `m`, `f`, `u` - undefined, `h` - hermaphrodite)

**Comment by goldbergtatyana** [comment]
are there gender-undefined pokemons?

**Comment by jonas-he** [comment]
Yes, the concept of gender was introduced in the second generation of pokemon, so technically every first generation pokemon (= Pokemon Go) except the Nidoran family has an undefined gender.

**Comment by samitsv** [comment]
but here it shows the gender of pokemons http://bulbapedia.bulbagarden.net/wiki/Gender

**Comment by goldbergtatyana** [comment]
yeah but then this link says "however, the genders of some species of Pokémon were left unknown". Btw, it would be a very nice statistics to show how many pokemons are m, f, h and u. :smile:

**Comment by jonas-he** [comment]
From what I see on the page you can only tell gender for: male: Nidoran♂, Nidorino, Nidoking, Hitmonlee, Hitmonchan, Tauros, female: Nidoran♀, Nidorina, Nidoqueen, Chansey, Kangaskhan and Jynx, the rest on the page is gen 2 or above => the rest of gen 1 is undefined. Will do changes accordingly if noone has something against it.

**Comment by goldbergtatyana** [comment]
The website below says that a pokemon can have a ratio of 1:7 to be a male (lol) and whether a pokemon is breedable.

Anyways, seems to be a good resource to look at.

http://bulbapedia.bulbagarden.net/wiki/List_of_Pok%C3%A9mon_by_gender_ratio

File: app/models/BasicPokemonDetail.js [hunk 5]
```diff
@@ -12,8 +12,7 @@ let BasicPokemonDetailSchema = new mongoose.Schema({
         type: String
     },
     gender: {
-        type: String
-        enum: ['m', 'f', 'u', 'h']
```
Comment (jonas-he) [inline app/models/BasicPokemonDetail.js]: Oh yeah, missed a comma, sorry ...

File: app/models/BasicPokemonDetail.js [hunk 6]
```diff
@@ -11,14 +12,7 @@ let BasicPokemonDetailSchema = new mongoose.Schema({
...
     gender: {
-        /*male: {
-            type: Boolean
-        },
-        female: {
-            type: Boolean
-        }*/
-        male: Boolean
-
+        type: String  //only use of values 'm' for male, 'f' for female, 'u' for undefined and 'h'
```
Comment (sacdallago) [inline app/models/BasicPokemonDetail.js]: There is a way to limit entries in a mognoose schema!! ;)

http://mongoosejs.com/docs/api.html#schema_string_SchemaString-enum

And I suggest you also normalize input by http://mongoosejs.com/docs/api.html#schema_string_SchemaString-lowercase

I don't though know the sequence in enum + toLowerCase (meaning, if it will first apply the lower case to, for example ,`M` and then validate it (which will be ok), or the other way around. The first way would work, the second way would produce an error. Test it! :smile: ).

Third option is write a validating function:
http://mongoosejs.com/docs/api.html#schematype_SchemaType-validate
